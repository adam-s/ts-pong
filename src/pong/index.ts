export * from './Pong';
export * from './Ball';
export * from './Court';
export * from './Menu';
export * from './Paddle';
export * from './resources';
