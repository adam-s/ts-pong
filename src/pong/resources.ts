import { ImageSource, AudioSource, Loader } from '../loader';

const loader = new Loader();

const images = {
  press1: new ImageSource('/assets/images/press1.png'),
  press2: new ImageSource('/assets/images/press2.png'),
  winner: new ImageSource('/assets/images/winner.png'),
};

for (const res in images) {
  loader.addResource(images[res as keyof typeof images]);
}

const sounds = {
  ping: new AudioSource('/assets/sounds/ping.wav'),
  pong: new AudioSource('/assets/sounds/pong.wav'),
  wall: new AudioSource('/assets/sounds/wall.wav'),
  goal: new AudioSource('/assets/sounds/goal.wav'),
};

for (const res in sounds) {
  const sound = sounds[res as keyof typeof sounds];
  sound.volume = 0.5;
  loader.addResource(sound);
}

export { images, sounds, loader };
