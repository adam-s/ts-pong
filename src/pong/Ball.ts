/* eslint-disable @typescript-eslint/no-empty-function */
import { Engine } from '../engine';
import { accelerate, Drawable, intercept, Updatable } from '../utils';
import { Paddle } from '../pong';
import { Colors } from './Pong';

export interface BallOptions {
  ballRadius: number;
  width: number;
  height: number;
  wallWidth: number;
  ballSpeed: number;
  ballAccel: number;
  footprints: boolean;
}

export class Ball implements Updatable, Drawable {
  private _x = 0;

  private _y = 0;

  private _dx: number;

  private _dy: number;

  private _radius: number;

  private minX: number;

  private maxX: number;

  private minY: number;

  private maxY: number;

  private speed: number;

  private accel: number;

  // private top: number;

  // private bottom: number;

  private _left: number;

  private _right: number;

  private dxChanged = false;

  private dyChanged = false;

  private footprints: Array<{ x: number; y: number }> = [];

  private footprintCount = 0;

  public showFootprints = false;

  constructor(options: BallOptions) {
    const {
      ballRadius,
      width,
      height,
      wallWidth,
      ballSpeed,
      ballAccel,
      footprints,
    } = options;
    this.minX = this._radius = ballRadius;
    this.maxX = width - this._radius;
    this.minY = wallWidth + this._radius;
    this.maxY = height - wallWidth - this._radius;
    this.speed = (this.maxX - this.minX) / ballSpeed;
    this.accel = ballAccel;
    this.showFootprints = footprints;
  }

  get x() {
    return this._x;
  }

  get y() {
    return this._y;
  }

  get dx() {
    return this._dx;
  }

  get dy() {
    return this._dy;
  }

  get radius() {
    return this._radius;
  }

  get left() {
    return this._left;
  }

  get right() {
    return this._right;
  }

  reset(playerNumber: number | null) {
    this.footprints = [];
    this.setpos(
      playerNumber === 1 ? this.maxX : this.minX,
      Engine.random(this.minY, this.maxY)
    );
    this.setdir(playerNumber === 1 ? -this.speed : this.speed, this.speed);
  }

  setpos(x: number, y: number) {
    this._x = x;
    this._y = y;
    this._left = this._x - this.radius;
    // this.top = this._y - this.radius;
    this._right = this._x + this.radius;
    // this.bottom = this._y + this.radius;
  }

  setdir(dx: number, dy: number) {
    this.dxChanged = this._dx < 0 !== dx < 0; // did horizontal direction change
    this.dyChanged = this._dy < 0 !== dy < 0; // did vertical direction change
    this._dx = dx;
    this._dy = dy;
  }

  footprint() {
    if (this.showFootprints) {
      if (!this.footprintCount || this.dxChanged || this.dyChanged) {
        this.footprints.push({ x: this.x, y: this.y });
        if (this.footprints.length > 50) this.footprints.shift();
        this.footprintCount = 5;
      } else {
        this.footprintCount--;
      }
    }
  }

  intercept(
    rect: { right: number; left: number; top: number; bottom: number },
    nx: number,
    ny: number
  ) {
    let pt;
    // Moving left
    if (nx < 0) {
      pt = intercept({
        x1: this.x, // x coord on the right side of ball segment
        y1: this.y, // y coord on the right side of ball segment
        x2: this.x + nx, // x coord where ball will be painted moving left
        y2: this.y + ny, // y coord where the ball will be painted moving left
        x3: rect.right + this.radius,
        y3: rect.top - this.radius,
        x4: rect.right + this.radius,
        y4: rect.bottom + this.radius,
        d: 'right',
      });
    } else if (nx > 0) {
      pt = intercept({
        x1: this.x,
        y1: this.y,
        x2: this.x + nx,
        y2: this.y + ny,
        x3: rect.left - this.radius,
        y3: rect.top - this.radius,
        x4: rect.left - this.radius,
        y4: rect.bottom + this.radius,
        d: 'left',
      });
    }
    if (!pt) {
      if (ny < 0) {
        pt = intercept({
          x1: this.x,
          y1: this.y,
          x2: this.x + nx,
          y2: this.y + ny,
          x3: rect.left - this.radius,
          y3: rect.bottom + this.radius,
          x4: rect.right + this.radius,
          y4: rect.bottom + this.radius,
          d: 'bottom',
        });
      } else if (ny > 0) {
        pt = intercept({
          x1: this.x,
          y1: this.y,
          x2: this.x + nx,
          y2: this.y + ny,
          x3: rect.left - this.radius,
          y3: rect.top - this.radius,
          x4: rect.right + this.radius,
          y4: rect.top - this.radius,
          d: 'top',
        });
      }
    }
    return pt;
  }

  update(dt: number, leftPaddle: Paddle, rightPaddle: Paddle) {
    const pos = accelerate({
      x: this.x,
      y: this.y,
      dx: this.dx,
      dy: this.dy,
      accel: this.accel,
      dt,
    });

    // Reverse direction (up) if ball hits bottom wall
    if (pos.dy > 0 && pos.y > this.maxY) {
      pos.y = this.maxY; // Set the y position to the top of the bottom wal
      pos.dy = -pos.dy; // Reverse direction
      // Reverse direction (down) if ball hits top wall
    } else if (pos.dy < 0 && pos.y < this.minY) {
      pos.y = this.minY; // Set the y position to the bottom of the top wall
      pos.dy = -pos.dy; // Reverse direction
    }

    // Choose paddle based on direction of ball. Negative from right to left is left paddle
    const paddle = pos.dx < 0 ? leftPaddle : rightPaddle;
    // Test if the ball will intercept the current paddle
    const pt = this.intercept(paddle, pos.nx, pos.ny);
    // If there is an interception with a paddle set the position to the point of intercept
    // Switch directions based on left / right or top / bottom
    if (pt) {
      switch (pt.d) {
        case 'left':
        case 'right':
          pos.x = pt.x;
          pos.dx = -pos.dx;
          break;
        case 'top':
        case 'bottom':
          pos.y = pt.y;
          pos.dy = -pos.dy;
          break;
      }

      // add/remove spin based on paddle direction
      if (paddle.up) pos.dy = pos.dy * (pos.dy < 0 ? 0.5 : 1.5);
      else if (paddle.down) pos.dy = pos.dy * (pos.dy > 0 ? 0.5 : 1.5);
    }

    this.setpos(pos.x, pos.y);
    this.setdir(pos.dx, pos.dy);
    this.footprint();
  }

  draw(ctx: CanvasRenderingContext2D) {
    const side = this.radius * 2;
    ctx.fillStyle = Colors.ball;
    ctx.fillRect(this.x - this.radius, this.y - this.radius, side, side);
    if (this.showFootprints) {
      const max = this.footprints.length;
      ctx.strokeStyle = Colors.footprint;
      for (let n = 0; n < max; n++)
        ctx.strokeRect(
          this.footprints[n].x - this.radius,
          this.footprints[n].y - this.radius,
          side,
          side
        );
    }
  }
}
