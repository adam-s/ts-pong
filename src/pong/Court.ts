import { Drawable, BlockElement } from '../utils';
import { Colors } from './Pong';

interface CourtOptions {
  width: number;
  height: number;
  wallWidth: number;
}

const DIGITS = [
  [1, 1, 1, 0, 1, 1, 1], // 0
  [0, 0, 1, 0, 0, 1, 0], // 1
  [1, 0, 1, 1, 1, 0, 1], // 2
  [1, 0, 1, 1, 0, 1, 1], // 3
  [0, 1, 1, 1, 0, 1, 0], // 4
  [1, 1, 0, 1, 0, 1, 1], // 5
  [1, 1, 0, 1, 1, 1, 1], // 6
  [1, 0, 1, 0, 0, 1, 0], // 7
  [1, 1, 1, 1, 1, 1, 1], // 8
  [1, 1, 1, 1, 0, 1, 0], // 9
] as const;

export class Court implements Drawable {
  private wallWidth: number;

  private walls: BlockElement[] = [];

  private score1: BlockElement;

  private score2: BlockElement;

  constructor(options: CourtOptions) {
    const { width, height, wallWidth } = options;
    this.wallWidth = wallWidth;

    // Top wall
    this.walls.push({ x: 0, y: 0, width, height: wallWidth });

    // Bottom wall
    this.walls.push({ x: 0, y: height - wallWidth, width, height: wallWidth });

    // Draw vertical dashed line halfway
    const innerHeight = height / (wallWidth * 2);
    for (let n = 0; n < innerHeight; n++) {
      this.walls.push({
        x: width / 2 - wallWidth / 2,
        y: wallWidth / 2 + wallWidth * 2 * n,
        width: wallWidth,
        height: wallWidth,
      });
    }

    const scoreWidth = 3 * wallWidth;
    const scoreHeight = 4 * wallWidth;

    this.score1 = {
      x: 0.5 + width / 2 - 1.5 * wallWidth - scoreWidth,
      y: 2 * wallWidth,
      width: scoreWidth,
      height: scoreHeight,
    };

    this.score2 = {
      x: 0.5 + width / 2 + 1.5 * wallWidth,
      y: 2 * wallWidth,
      width: scoreWidth,
      height: scoreHeight,
    };
  }

  draw(
    ctx: CanvasRenderingContext2D,
    scorePlayer1: number,
    scorePlayer2: number
  ) {
    ctx.fillStyle = Colors.walls;
    for (let i = 0; i < this.walls.length; i++) {
      const { x, y, width, height } = this.walls[i];
      ctx.fillRect(x, y, width, height);
    }

    this.drawDigit(ctx, scorePlayer1, this.score1);
    this.drawDigit(ctx, scorePlayer2, this.score2);
  }

  drawDigit(
    ctx: CanvasRenderingContext2D,
    score: number,
    scoreElement: BlockElement
  ) {
    const { x, y, width, height } = scoreElement;
    ctx.fillStyle = Colors.walls;

    const drawWidth = (this.wallWidth * 4) / 5;
    const drawHeight = drawWidth; // For self documentation

    const blocks = DIGITS[score];

    if (blocks[0]) {
      ctx.fillRect(x, y, width, drawWidth);
    }
    if (blocks[1]) {
      ctx.fillRect(x, y, drawWidth, height / 2);
    }
    if (blocks[2]) {
      ctx.fillRect(x + width - drawWidth, y, drawWidth, height / 2);
    }
    if (blocks[3]) {
      ctx.fillRect(x, y + height / 2 - drawHeight / 2, width, drawHeight);
    }
    if (blocks[4]) {
      ctx.fillRect(x, y + height / 2, drawWidth, height / 2);
    }
    if (blocks[5]) {
      ctx.fillRect(
        x + width - drawWidth,
        y + height / 2,
        drawWidth,
        height / 2
      );
    }
    if (blocks[6]) {
      ctx.fillRect(x, y + height - drawHeight, width, drawHeight);
    }
  }
}
