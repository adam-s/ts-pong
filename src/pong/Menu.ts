import { Drawable, ImageElement } from '../utils';
import { PlayerNumber } from './Pong';

interface MenuOptions {
  images: { [key: string]: HTMLImageElement };
  width: number;
  wallWidth: number;
}

export class Menu implements Drawable {
  private press1: ImageElement;

  private press2: ImageElement;

  private winner1: ImageElement;

  private winner2: ImageElement;

  private winner: PlayerNumber;

  constructor(options: MenuOptions) {
    const { images, width, wallWidth } = options;
    const press1 = images['press1'];
    const press2 = images['press2'];
    const winner = images['winner'];
    this.press1 = { image: press1, x: 10, y: wallWidth };
    this.press2 = {
      image: press2,
      x: width - press2.width - 10,
      y: wallWidth,
    };
    this.winner1 = {
      image: winner,
      x: width / 2 - winner.width - wallWidth,
      y: 6 * wallWidth,
    };
    this.winner2 = {
      image: winner,
      x: width / 2 + wallWidth,
      y: 6 * wallWidth,
    };
  }

  declareWinner(playerNumber: PlayerNumber) {
    this.winner = playerNumber;
  }

  draw(ctx: CanvasRenderingContext2D) {
    // Maybe wait for images to load
    ctx.drawImage(this.press1.image, this.press1.x, this.press1.y);
    ctx.drawImage(this.press2.image, this.press2.x, this.press2.y);
    if (this.winner == 0)
      ctx.drawImage(this.winner1.image, this.winner1.x, this.winner1.y);
    else if (this.winner == 1)
      ctx.drawImage(this.winner2.image, this.winner2.x, this.winner2.y);
  }
}
