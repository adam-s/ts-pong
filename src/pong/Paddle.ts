import { Colors, Level, Levels } from './Pong';
import { Direction, Drawable, Updatable } from '../utils';
import { Ball } from './Ball';
import { Engine } from '../engine/Engine';

/* eslint-disable @typescript-eslint/no-empty-function */
interface PaddleOptions {
  rightHandSide?: boolean;
  paddleWidth: number;
  paddleHeight: number;
  paddleSpeed: number;
  wallWidth: number;
  height: number;
  width: number;
  showPredictions: boolean;
}

interface Predication {
  x: number;
  y: number;
  d: Direction;
  dx: number;
  dy: number;
  exactX: number;
  exactY: number;
  radius: number;
  since: number;
}

export class Paddle implements Drawable, Updatable {
  private width: number;

  private height: number;

  private paddleWidth: number;

  private paddleHeight: number;

  private minY: number;

  private maxY: number;

  private speed: number;

  private x: number;

  private y: number;

  private _left: number;

  private _right: number;

  private _top: number;

  private _bottom: number;

  private _up: number;

  private _down: number;

  private _auto = false;

  private level: Level;

  private prediction: Predication;

  public showPredications: boolean;

  constructor(options: PaddleOptions) {
    const {
      rightHandSide,
      paddleWidth,
      paddleHeight,
      paddleSpeed,
      wallWidth,
      height,
      width,
      showPredictions,
    } = options;

    this.width = width;
    this.height = height;
    this.paddleWidth = paddleWidth;
    this.paddleHeight = paddleHeight;
    this.minY = wallWidth;
    this.maxY = height - wallWidth - this.paddleHeight;
    this.speed = (this.maxY - this.minY) / paddleSpeed;
    this.setpos(
      rightHandSide ? width - paddleWidth : 0,
      this.minY + (this.maxY - this.minY) / 2
    );
    this.setdir(0);
    this.level = Levels[8];
    this.showPredications = showPredictions;
  }

  setpos(x: number, y: number) {
    this.x = x;
    this.y = y;
    this._left = this.x;
    this._right = this._left + this.paddleWidth;
    this._top = this.y;
    this._bottom = this.y + this.paddleHeight;
  }

  setdir(dy: number) {
    this._up = dy < 0 ? -dy : 0;
    this._down = dy > 0 ? dy : 0;
  }

  setAuto(on: boolean): void;
  setAuto(on: boolean, level: number): void;
  setAuto(on: boolean, level?: number) {
    if (on && !this._auto && level) {
      this._auto = true;
      this.setLevel(level);
    } else if (!on && this._auto) {
      this._auto = false;
      this.setdir(0);
    }
  }

  get right() {
    return this._right;
  }

  get left() {
    return this._left;
  }

  get top() {
    return this._top;
  }

  get bottom() {
    return this._bottom;
  }

  get up() {
    return this._up;
  }

  get down() {
    return this._down;
  }

  get auto() {
    return this._auto;
  }

  setLevel(level: number) {
    if (this._auto) this.level = Levels[level];
  }

  update(dt: number, ball: Ball) {
    if (this._auto) {
      this.ai(dt, ball);
    }

    const amount = this._down - this._up;
    if (amount !== 0) {
      let y = this.y + amount * dt * this.speed;
      if (y < this.minY) {
        y = this.minY;
      } else if (y > this.maxY) {
        y = this.maxY;
      }
      this.setpos(this.x, y);
    }
  }

  ai(dt: number, ball: Ball) {
    if (
      (ball.x < this._left && ball.dx < 0) ||
      (ball.x > this._right && ball.dx > 0)
    ) {
      this.stopMovingUp();
      this.stopMovingDown();
    } else {
      this.predict(dt, ball);
      if (this.prediction) {
        if (this.prediction.y < this._top + this.paddleHeight / 2 - 5) {
          this.stopMovingDown();
          this.moveUp();
        } else if (
          this.prediction.y >
          this._bottom - this.paddleHeight / 2 + 5
        ) {
          this.stopMovingUp();
          this.moveDown();
        } else {
          this.stopMovingUp();
          this.stopMovingDown();
        }
      }
    }
  }

  predict(dt: number, ball: Ball) {
    // only re-predict if the ball changed direction, or its been some amount of time since last prediction
    if (
      this.prediction &&
      this.prediction.dx * ball.dx > 0 &&
      this.prediction.dy * ball.dy > 0 &&
      this.prediction.since < this.level.aiReaction
    ) {
      this.prediction.since += dt;
    } else {
      const pt = ball.intercept(
        {
          left: this._left,
          right: this._right,
          top: -10000,
          bottom: 10000,
        },
        ball.dx * 10,
        ball.dy * 10
      );

      if (pt) {
        const t = this.minY + ball.radius;
        const b = this.maxY + this.height - ball.radius;

        while (pt.y < t || pt.y > b) {
          if (pt.y < t) {
            pt.y = t + (t - pt.y);
          } else if (pt.y > b) {
            pt.y = t + (b - t) - (pt.y - b);
          }
        }
        this.prediction = {
          since: 0,
          dx: ball.dx,
          dy: ball.dy,
          x: pt.x,
          y: pt.y,
          radius: ball.radius,
          exactX: pt.x,
          exactY: pt.y,
          d: pt.d,
        };
        const closeness =
          (ball.dx < 0 ? ball.x - this._right : this._left - ball.x) /
          this.width;
        const error = this.level.aiError * closeness;
        this.prediction.y = this.prediction.y + Engine.random(-error, error);
      } else {
        this.prediction = null;
      }
    }
  }

  moveUp() {
    this._up = 1;
  }

  moveDown() {
    this._down = 1;
  }

  stopMovingUp() {
    this._up = 0;
  }

  stopMovingDown() {
    this._down = 0;
  }

  draw(ctx: CanvasRenderingContext2D) {
    ctx.fillStyle = Colors.walls;
    ctx.fillRect(this.x, this.y, this.paddleWidth, this.paddleHeight);
    if (this.prediction && this.showPredications) {
      ctx.strokeStyle = Colors.predictionExact;
      ctx.strokeRect(
        this.prediction.x - this.prediction.radius,
        this.prediction.exactY - this.prediction.radius,
        this.prediction.radius * 2,
        this.prediction.radius * 2
      );
      ctx.strokeStyle = Colors.predictionGuess;
      ctx.strokeRect(
        this.prediction.x - this.prediction.radius,
        this.prediction.y - this.prediction.radius,
        this.prediction.radius * 2,
        this.prediction.radius * 2
      );
    }
  }
}
