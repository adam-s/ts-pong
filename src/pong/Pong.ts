/* eslint-disable @typescript-eslint/no-empty-function */
import { Game } from '../engine/Engine';
import { Keys } from '../utils';
import { Ball } from './Ball';
import { Court } from './Court';
import { Menu } from './Menu';
import { Paddle } from './Paddle';
import { images, sounds } from './resources';

export type PlayerNumber = 0 | 1;

export type Scores = [number, number];

export type Level = { readonly aiReaction: number; readonly aiError: number };

export const Config = {
  width: 640, // logical canvas width (browser will scale to physical canvas size - which is controlled by @media css queries)
  height: 480, // logical canvas height (ditto)
  wallWidth: 12,
  paddleWidth: 12,
  paddleHeight: 60,
  paddleSpeed: 2, // should be able to cross court vertically   in 2 seconds
  ballSpeed: 4, // should be able to cross court horizontally in 4 seconds, at starting speed ...
  ballAccel: 8, // ... but accelerate as time passes
  ballRadius: 5,
  playSounds: true,
  showFootprints: false,
  showPredications: false,
} as const;

export const Colors = {
  walls: 'white',
  ball: 'white',
  score: 'white',
  footprint: '#333',
  predictionGuess: 'yellow',
  predictionExact: 'red',
} as const;

export const Levels = [
  { aiReaction: 0.2, aiError: 40 }, // 0:  ai is losing by 8
  { aiReaction: 0.3, aiError: 50 }, // 1:  ai is losing by 7
  { aiReaction: 0.4, aiError: 60 }, // 2:  ai is losing by 6
  { aiReaction: 0.5, aiError: 70 }, // 3:  ai is losing by 5
  { aiReaction: 0.6, aiError: 80 }, // 4:  ai is losing by 4
  { aiReaction: 0.7, aiError: 90 }, // 5:  ai is losing by 3
  { aiReaction: 0.8, aiError: 100 }, // 6:  ai is losing by 2
  { aiReaction: 0.9, aiError: 110 }, // 7:  ai is losing by 1
  { aiReaction: 1.0, aiError: 120 }, // 8:  tie
  { aiReaction: 1.1, aiError: 130 }, // 9:  ai is winning by 1
  { aiReaction: 1.2, aiError: 140 }, // 10: ai is winning by 2
  { aiReaction: 1.3, aiError: 150 }, // 11: ai is winning by 3
  { aiReaction: 1.4, aiError: 160 }, // 12: ai is winning by 4
  { aiReaction: 1.5, aiError: 170 }, // 13: ai is winning by 5
  { aiReaction: 1.6, aiError: 180 }, // 14: ai is winning by 6
  { aiReaction: 1.7, aiError: 190 }, // 15: ai is winning by 7
  { aiReaction: 1.8, aiError: 200 }, // 16: ai is winning by 8
] as const;

export class Pong implements Game {
  public readonly width = Config.width;

  public readonly height = Config.height;

  private _scores: Scores = [0, 0];

  private _playing = false;

  private _menu: Menu;

  private _court: Court;

  private _leftPaddle: Paddle;

  private _rightPaddle: Paddle;

  private _ball: Ball;

  private _images: { [key: string]: HTMLImageElement } = {};

  private _playSounds = false;

  constructor() {
    this._playSounds = Config.playSounds;

    const paddleOptions = {
      paddleWidth: Config.paddleWidth,
      paddleHeight: Config.paddleHeight,
      paddleSpeed: Config.paddleSpeed,
      wallWidth: Config.wallWidth,
      height: Config.height,
      width: Config.width,
      showPredictions: Config.showPredications,
    };
    this._leftPaddle = new Paddle(
      Object.assign({}, paddleOptions, {
        rightHandSide: false,
      })
    );
    this._rightPaddle = new Paddle(
      Object.assign({}, paddleOptions, {
        rightHandSide: true,
      })
    );
    this._ball = new Ball({
      ballRadius: Config.ballRadius,
      width: Config.width,
      height: Config.height,
      wallWidth: Config.wallWidth,
      ballSpeed: Config.ballSpeed,
      ballAccel: Config.ballAccel,
      footprints: Config.showFootprints,
    });

    this._court = new Court({
      width: Config.width,
      height: Config.height,
      wallWidth: Config.wallWidth,
    });

    // The menu requires loading images first so this has to be called in
    // an async init() method ðŸ˜¡ I'm and idiot.
    const promises = Object.values(images).map((image) => image.ready);
    Promise.all(promises).then(() => {
      for (const image in images) {
        this._images[image] = images[image as keyof typeof images].image;
      }
      this._menu = new Menu({
        images: this._images,
        width: Config.width,
        wallWidth: Config.wallWidth,
      });
    });
  }

  startDemo() {
    this.start(0);
  }
  startSinglePlayer() {
    this.start(1);
  }
  startDoublePlayer() {
    this.start(2);
  }

  start(numPlayers: 0 | 1 | 2) {
    if (!this._playing) {
      this._scores = [0, 0];
      this._playing = true;
      this._leftPaddle.setAuto(numPlayers < 1, this.level(0));
      this._rightPaddle.setAuto(numPlayers < 2, this.level(1));
      this._ball.reset(null);
      // this._runner.hideCursor();
    }
  }

  stop(ask?: boolean): void {
    ask;
    if (this._playing) {
      // if (!ask || this.runner.confirm('Abandon game in progress ?')) {
      this._playing = false;
      this._leftPaddle.setAuto(false);
      this._rightPaddle.setAuto(false);
      // this.runner.showCursor();
      // }
    }
  }

  level(playerNumber: PlayerNumber): number {
    return (
      8 + (this._scores[playerNumber] - this._scores[playerNumber ? 0 : 1])
    );
  }

  sounds(sound: keyof typeof sounds) {
    if (this._playSounds) {
      switch (sound) {
        case 'goal':
          sounds.goal.play();
          break;
        case 'ping':
          sounds.ping.play();
          break;
        case 'pong':
          sounds.pong.play();
          break;
        case 'wall':
          sounds.wall.play();
          break;
      }
    }
  }

  goal(playerNumber: PlayerNumber) {
    this.sounds('goal');
    this._scores[playerNumber] += 1;
    if (this._scores[playerNumber] == 9) {
      this._menu.declareWinner(playerNumber);
      this.stop();
    } else {
      this._ball.reset(playerNumber);
      this._leftPaddle.setLevel(this.level(0));
      this._rightPaddle.setLevel(this.level(1));
    }
  }

  update(dt: number): void {
    this._leftPaddle.update(dt, this._ball);
    this._rightPaddle.update(dt, this._ball);
    if (this._playing) {
      // Get the previous position
      const dx = this._ball.dx;
      const dy = this._ball.dy;

      // Update the position
      this._ball.update(dt, this._leftPaddle, this._rightPaddle);

      // Sounds based on difference between previous and current updated positions
      if (this._ball.dx < 0 && dx > 0) {
        this.sounds('ping');
      } else if (this._ball.dx > 0 && dx < 0) {
        this.sounds('pong');
      } else if (this._ball.dy * dy < 0) {
        this.sounds('wall');
      }

      if (this._ball.left > this.width) this.goal(0);
      else if (this._ball.right < 0) this.goal(1);
    }
  }

  draw(ctx: CanvasRenderingContext2D): void {
    this._court.draw(ctx, this._scores[0], this._scores[1]);
    this._leftPaddle.draw(ctx);
    this._rightPaddle.draw(ctx);
    if (this._playing) this._ball.draw(ctx);
    else this._menu.draw(ctx);
  }

  onkeydown(ev: KeyboardEvent) {
    switch (ev.code) {
      case Keys.Key0:
        this.startDemo();
        break;
      case Keys.Key1:
        this.startSinglePlayer();
        break;
      case Keys.Key2:
        this.startDoublePlayer();
        break;
      case Keys.Escape:
        this.stop(true);
        break;
      case Keys.Q:
        if (!this._leftPaddle.auto) this._leftPaddle.moveUp();
        break;
      case Keys.A:
        if (!this._leftPaddle.auto) this._leftPaddle.moveDown();
        break;
      case Keys.P:
        if (!this._rightPaddle.auto) this._rightPaddle.moveUp();
        break;
      case Keys.L:
        if (!this._rightPaddle.auto) this._rightPaddle.moveDown();
        break;
    }
  }

  onkeyup(ev: KeyboardEvent) {
    switch (ev.code) {
      case Keys.Q:
        if (!this._leftPaddle.auto) this._leftPaddle.stopMovingUp();
        break;
      case Keys.A:
        if (!this._leftPaddle.auto) this._leftPaddle.stopMovingDown();
        break;
      case Keys.P:
        if (!this._rightPaddle.auto) this._rightPaddle.stopMovingUp();
        break;
      case Keys.L:
        if (!this._rightPaddle.auto) this._rightPaddle.stopMovingDown();
        break;
    }
  }

  showFootprints(show: boolean) {
    // Have to toggle showFootprints on ball
    this._ball.showFootprints = show;
  }

  showPredictions(show: boolean) {
    // Have to toggle showPredications on the paddle
    this._leftPaddle.showPredications = show;
    this._rightPaddle.showPredications = show;
  }

  playSounds(enabled: boolean) {
    // Have to toggle enableSound on this (pong)
    this._playSounds = enabled;
  }
}
