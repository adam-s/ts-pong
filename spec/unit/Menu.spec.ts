import 'jest-canvas-mock';
import { CanvasRenderingContext2DEvent } from 'jest-canvas-mock';
import { Menu } from '../../src/pong/Menu';

const mockImage = (width: number, height: number) => {
  Object.defineProperties(Image.prototype, {
    width: { value: width },
    height: { value: height },
    src: {
      set: (src: string) => {
        console.log('src', src);
      },
    },
  });
  const image = new Image();
  return image;
};

describe('Menu', () => {
  it('should exist', () => {
    expect(Menu).toBeDefined();
  });
  it('draw a menu', () => {
    const images = {
      'images/press1.png': mockImage(1, 1),
      'images/press2.png': mockImage(2, 2),
      'images/winner.png': mockImage(3, 3),
    };
    let events: CanvasRenderingContext2DEvent[];
    const implementation = new Menu({ images, width: 100, wallWidth: 100 });
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d');

    // No winner
    implementation.draw(ctx);
    events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
    ctx.__clearEvents();

    // Player one is winner
    implementation.declareWinner(0);
    implementation.draw(ctx);
    events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
    ctx.__clearEvents();

    // Player two is winner
    implementation.declareWinner(1);
    implementation.draw(ctx);
    events = ctx.__getEvents();
    expect(events).toMatchSnapshot();
    ctx.__clearEvents();
  });
});
