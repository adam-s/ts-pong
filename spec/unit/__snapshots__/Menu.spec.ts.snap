// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu draw a menu 1`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": 10,
      "dy": 100,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": 87,
      "dy": 100,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`Menu draw a menu 2`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": 10,
      "dy": 100,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": 87,
      "dy": 100,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": -53,
      "dy": 600,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`Menu draw a menu 3`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": 10,
      "dy": 100,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": 87,
      "dy": 100,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 3,
      "dWidth": 3,
      "dx": 150,
      "dy": 600,
      "img": <img />,
      "sHeight": 3,
      "sWidth": 3,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
